### PR TITLE
CELDEV-1076 - bugfixing for tinymce4

### PR DESCRIPTION
**Pull request by fpichler**
https://synjira.atlassian.net/browse/CELDEV-1076

1. Fixes for loading javascripts lazyly in order of mentioning.
2. Fixes for contextmenu e.g. beeing displayed outside of window

[CELDEV-1076]: https://synjira.atlassian.net/browse/CELDEV-1076?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/webapp/resources/celDynJS/DynamicLoader/celLazyLoader.mjs [hunk 1]
```diff
@@ -102,9 +104,41 @@ class CelLazyLoaderUtils {
...
+      console.debug('_loadScriptElem insert ', newEle);
+      customElem._reayState = 1;
+      if (loadMode && (loadMode === 'async')) {
+        document.head.appendChild(newEle);
+      } else {
+        const lastPromise = lastPromiseOfLoadingType[elemType] ?? Promise.resolve();
+        lastPromiseOfLoadingType[elemType] = lastPromise.then(() => {
+          document.head.appendChild(newEle);
+          return loadedPromise;
+        });
```
Comment (msladek) [inline src/main/webapp/resources/celDynJS/DynamicLoader/celLazyLoader.mjs]:
```suggestion
        await lastPromise;
        document.head.appendChild(newEle);
        lastPromiseOfLoadingType[elemType] = loadedPromise;
```
NOTE: async needed in function def

File: src/main/webapp/resources/celJS/contextMenu/contextmenu.js [hunk 2]
```diff
@@ -19,61 +19,50 @@
...
+};
 
-function contextMouseOut(n){
-  classAttribute = n.getAttributeNode('class');
+const contextMouseOut = function(n) {
+  const classAttribute = n.getAttributeNode('class');
   classAttribute.nodeValue = "contextMenuLink";
   
-  $(document).observe('mousedown', myContextMenu.hide);
+  document.addEventListener('mousedown', myContextMenu.hideBind);
```
Comment (msladek) [inline src/main/webapp/resources/celJS/contextMenu/contextmenu.js]: the bind `hideBind` shouldn't be necessary with arrow functions:
```suggestion
  document.addEventListener('mousedown', e => myContextMenu.hide(e));
```
can be replaced everywhere like so

File: src/main/webapp/resources/celJS/contextMenu/contextmenu.js [hunk 3]
```diff
@@ -100,90 +89,89 @@ function ContextMenuItem(link, text, icon, shortcut) {
...
-  
-  this.setPosition = function (y, x) {
-    me.menuDiv.setStyle({
+  }
+
+  setPosition(y, x) {
+    this.getMenuDiv().setStyle({
       'left' : x + 'px',
       'top' : y + 'px'
     });
```
Comment (msladek) [inline src/main/webapp/resources/celJS/contextMenu/contextmenu.js]:
```suggestion
    this.getMenuDiv().style.left = x + 'px';
    this.getMenuDiv().style.top = y + 'px';
```

File: src/main/webapp/resources/celJS/contextMenu/contextmenu.js [hunk 4]
```diff
@@ -100,90 +89,89 @@ function ContextMenuItem(link, text, icon, shortcut) {
...
+  #menuDiv;
+  hideBind;
+  
+  constructor() {
+    this.#config = [];
+    this.hideBind = this.hide.bind(this);
+  }
 
-function ContextMenu(){
+  getMenuDiv() {
```
Comment (msladek) [inline src/main/webapp/resources/celJS/contextMenu/contextmenu.js]: you could define a [class getter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get#using_getters_in_classes) instead
```suggestion
  get menuDiv() {
```
and keep the previous usage of `this.menuDiv`

File: src/main/webapp/resources/celJS/contextMenu/contextmenu.js [hunk 5]
```diff
@@ -100,90 +89,89 @@ function ContextMenuItem(link, text, icon, shortcut) {
...
-      me.menuDiv.remove();
-    me.menuDiv = null;
+  }
+
+  internal_hide() {
+    this.#config = [];
+    if (this.#menuDiv) {
+      this.#menuDiv.remove();
+    }
+    this.#menuDiv = undefined;
```
Comment (msladek) [inline src/main/webapp/resources/celJS/contextMenu/contextmenu.js]:
```suggestion
    this.#menuDiv = null;
```

File: src/main/webapp/templates/celMacros/loadAdminInterface.vm [hunk 6]
```diff
@@ -2,5 +2,36 @@
...
+  #end
+##   $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/validation.js') ## always loaded in includeDefaultHeader.vm
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/adminUi/menubar.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/supportLink/supportLink.js'))
+## keep the following if async with the one in celMacros/startCelementsBody.vm to load getTabMenu!!
+  #if($cel_isEditMode && !$cel_useInlineEditorMode)
+    #parse("celMacros/DarkBox.vm")
+    $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/celTabMenu/TabEditor.js'))
+    $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/celTabMenu/celementsTabMenu.js'))
+    $celementsweb.includeCSSPage(':celRes/celTabMenu/tabMenu.css')
```
Comment (msladek) [inline src/main/webapp/templates/celMacros/loadAdminInterface.vm]:
```suggestion
    $!services.css.includeCSSPage(':celRes/celTabMenu/tabMenu.css')
```

File: src/main/webapp/templates/celMacros/loadAdminInterface.vm [hunk 7]
```diff
@@ -2,5 +2,36 @@
...
+
+#if($hasedit || $isContentEditor || $isAdmin || ($services.celMenu.getMenuHeaders().size() > 0))
+  #set($jsService = $services.javascript)
+  #set($jsLoadModeDefer = $!jsService.getJsLoadMode("DEFER"))
+  #set($extJsParamDefer = $!jsService.createExtJSParam().setAction('file').setLoadMode($jsLoadModeDefer))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/moofx/moo.fx.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/moofx/moo.fx.utils.js'))
+##  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/adminUi/div.js'))
+##  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/adminUi/aspects.js'))
+##  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/adminUi/window.js'))
```
Comment (msladek) [inline src/main/webapp/templates/celMacros/loadAdminInterface.vm]: why commented out ? explain in the comment

**Review by msladek — CHANGES_REQUESTED** [review]

File: src/main/webapp/templates/celMacros/loadAdminInterface.vm (hunk 7 above)
Comment (fpichler) [reply src/main/webapp/templates/celMacros/loadAdminInterface.vm]: It was already commented out before inlining. As it was commented out for about 10 years, I just removed now :-)

**Review by msladek — APPROVED** [review]

**Pull request merged**